tsl/src/remote/async.c

```diff
@@ -512,6 +512,137 @@ async_request_wait_ok_result(AsyncRequest *req)
 	return res;
 }
 
+/*
+ * Get the result of an async request during cleanup.
+ *
+ * Cleanup is typically necessary for a query that is being interrupted by
+ * transaction abort, or a query that was initiated as part of transaction
+ * abort to get the remote side back to the appropriate state.
+ *
+ * endtime is the time at which we should give up and assume the remote
+ * side is dead.
+ *
+ * An AsyncReponse is always returned, indicating last PGresult received,
+ * a timeout, or error.
+ */
+AsyncResponse *
+async_request_cleanup_result(AsyncRequest *req, TimestampTz endtime)
+{
+	volatile bool timed_out = false;
+	volatile bool conn_issue = false;
+	PGresult *volatile last_res = NULL;
+	const TSConnection *conn = async_request_get_connection(req);
+	PGconn *pg_conn = remote_connection_get_pg_conn(conn);
+
+	switch (req->state)
+	{
+		case DEFERRED:
+			if (remote_connection_is_processing(req->conn))
+				return async_response_error_create("request already in progress");
+
+			req = async_request_send_internal(req, WARNING);
+
+			if (req == NULL)
+				return async_response_error_create("failed to send deferred request");
+
+			Assert(req->state == EXECUTING);
+			break;
+		case EXECUTING:
+			break;
+		case COMPLETED:
+			return async_response_error_create("request already completed");
+	}
+
+	/* In what follows, do not leak any PGresults on an error. */
+	PG_TRY();
+	{
+		for (;;)
+		{
+			PGresult *res;
+
+			while (PQisBusy(pg_conn))
+			{
+				int wc;
+				TimestampTz now = GetCurrentTimestamp();
+				long secs;
+				int microsecs;
+				long cur_timeout;
+
+				/* If timeout has expired, give up, else get sleep time. */
+				if (now >= endtime)
+				{
+					timed_out = true;
+					goto exit;
+				}
+				TimestampDifference(now, endtime, &secs, &microsecs);
+
+				/* To protect against clock skew, limit sleep to one minute. */
+				cur_timeout = Min(60000, secs * USECS_PER_SEC + microsecs);
+
+				/* Sleep until there's something to do */
+				wc = WaitLatchOrSocket(MyLatch,
+									   WL_LATCH_SET | WL_SOCKET_READABLE
+#if PG12_GE
+										   | WL_EXIT_ON_PM_DEATH
+#endif
+										   | WL_TIMEOUT,
```

Comment:
It might be good to extract this flag combination into a constant or function as it seems to be similar to the flag combination in https://github.com/timescale/timescaledb/blob/master/tsl/src/remote/connection.c#L1524-L1525

Reply:
Second that. Even if just constants in this function is is a lot easier to read, especially considering that they are different between PG versions.

Reply:
I actually realized I could merge the code paths since they were based on the same original function. Now the code should be cleaner and the flags only appear in one place.